env/composite_env_wrapper.h

```diff
@@ -0,0 +1,1230 @@
+// be agnostic of whether the underlying Env implementation is a monolithic
+// Env or an Env + FileSystem. In the former case, the user will specify
+// Options::env only, whereas in the latter case, the user will specify
+// Options::env and Options::file_system.
+#define IO_STATUS_TO_STATUS(_status_, _io_status_)                 \
+  {                                                                \
+    IOStatus io_status = _io_status_;                              \
+    if (!io_status.ok()) {                                         \
+      _status_ = io_status;                                        \
+      if (io_status.GetRetryable()) {                                 \
+        _status_ = Status(_status_, Status::Severity::kSoftError); \
+      }                                                            \
+    } else {                                                       \
+      _status_ = Status::OK();                                     \
+    }                                                              \
+  }
+
+inline IOStatus status_to_io_status(Status&& status) {
+  if (status.ok()) {
+    // Fast path
+    return IOStatus::OK();
+  } else {
+    const char* state = status.getState();
+    return IOStatus(status.code(), status.subcode(),
+                    Slice(state, state ? strlen(status.getState()) + 1 : 0),
+                    Slice());
+  }
+}
+
+class CompositeSequentialFileWrapper : public SequentialFile {
+ public:
+  explicit CompositeSequentialFileWrapper(
+      std::unique_ptr<FSSequentialFile>& target)
+      : target_(std::move(target)) {}
+
+  Status Read(size_t n, Slice* result, char* scratch) override {
+    IOOptions io_opts;
+    IODebugContext dbg;
+    Status status;
+    IO_STATUS_TO_STATUS(status,
+                        target_->Read(n, io_opts, result, scratch, &dbg));
+    return status;
+  }
+  Status Skip(uint64_t n) override {
+    Status status;
+    IO_STATUS_TO_STATUS(status, target_->Skip(n));
+    return status;
+  }
+  bool use_direct_io() const override { return target_->use_direct_io(); }
+  size_t GetRequiredBufferAlignment() const override {
+    return target_->GetRequiredBufferAlignment();
+  }
+  Status InvalidateCache(size_t offset, size_t length) override {
+    Status status;
+    IO_STATUS_TO_STATUS(status, target_->InvalidateCache(offset, length));
+    return status;
+  }
+  Status PositionedRead(uint64_t offset, size_t n, Slice* result,
+                        char* scratch) override {
+    IOOptions io_opts;
+    IODebugContext dbg;
+    Status status;
+    IO_STATUS_TO_STATUS(status, target_->PositionedRead(offset, n, io_opts,
+                                                        result, scratch, &dbg));
+    return status;
+  }
+
+ private:
+  std::unique_ptr<FSSequentialFile> target_;
+};
+
+class CompositeRandomAccessFileWrapper : public RandomAccessFile {
+ public:
+  explicit CompositeRandomAccessFileWrapper(
+      std::unique_ptr<FSRandomAccessFile>& target)
```

Comment:
Should we use rvalue ref here?

Reply:
There are cases where we need to bind an lvalue to the argument, so rvalue ref won't work. For example, in ```CompositeEnvWrapper::NewRandomAccessFile()```.